### PR TITLE
consortium-v2/state: upgrade Axie proxy contract

### PR DESCRIPTION
**Pull request by andicrypt**
**Aaron** hardfork draft: 
- [x] Upgrade Axie and Land proxy code to achieve transparent proxy.   

- [x] Introduce block producer bit set in header's extra to replace the list of block producers.

   Block producer set is a subset of validator candidate set, but currently, both set are stored in plain, which costs unnecessary resource. This commit introduces BlockProducersBitSet (uint64) as an encoded representation of block producer list, indexing to validator candidate list. The original block producer set can be reconstructed from the bit set and validator candidate set. As a result, instead of storing a list of block producers in header extra, only a single bit set of type uint64 is stored.

### CODE REVIEW

**Comment by huyngopt1994** [comment]
@andicrypt Could we put unitest here ?

**Comment by andicrypt** [comment]
> @andicrypt Could we put unitest here ?

Do you mean a new Ronin-test suties for these changes? I just added some unit test for these changes.

File: params/config.go [hunk 1]
```diff
@@ -575,13 +578,19 @@ type ChainConfig struct {
...
 }
 
 type ContractUpgrade struct {
 	ProxyAddress          common.Address `json:"proxyAddress"`
 	ImplementationAddress common.Address `json:"implementationAddress"`
 }
 
+type ContractCodeUpgrade struct {
+	CodeAddress common.Address `json:"contractAddress"`
+	Code        []byte         `json:"code"`
```
Comment (minh-bq) [inline params/config.go:591]: https://pkg.go.dev/encoding/json#Marshal

Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON value.

So in the genesis.json, the code field must be base64 of real code. Is this an expected behavior?

File: consensus/consortium/v2/consortium_test.go [hunk 2]
```diff
@@ -1825,6 +1825,9 @@ func TestUpgradeRoninTrustedOrg(t *testing.T) {
 		Consortium: &params.ConsortiumConfig{
 			EpochV2: 200,
 		},
+		ConsortiumV2Contracts: &params.ConsortiumV2Contracts{
```
Comment (minh-bq) [inline consensus/consortium/v2/consortium_test.go:1828]: Why changing this?

File: params/config.go (hunk 1 above)
Comment (andicrypt) [reply params/config.go]: You are correct. I have changed the type of Code to string (hex), which will be loaded from genesis file, before being decoded into []byte for statedb process.

File: consensus/consortium/v2/consortium.go [hunk 3]
```diff
@@ -837,11 +848,23 @@ func (c *Consortium) Prepare(chain consensus.ChainHeaderReader, header *types.He
...
 		if c.IsTrippEffective(chain, header) {
 			if c.IsPeriodBlock(chain, header) {
-				extraData.CheckpointValidators = checkpointValidator
+				extraData.CheckpointValidators = checkpointValidators
+			} else {
+				// checkpointValidators get from contract is nil at epoch block; hence,
+				// checkpointValidators is reassigned by the candidate list from snapshot,
+				// which is used in encodingValidatorBitSet step.
+				checkpointValidators = snap.ValidatorsWithBlsPub
+			}
```
Comment (minh-bq) [inline consensus/consortium/v2/consortium.go:857]: I've just realized this is incorrect, we must always use snap.ValidatorsWithBlsPub. At period block, we don't switch to new block producer set right way but wait until len(blockProducer) / 2 (logic in snapshot). So the block producer here must be the indices of block producers in the old set in snap.ValidatorsWithBlsPub not the new set in extradata.

Comment (minh-bq) [reply consensus/consortium/v2/consortium.go:857]: This is an incorrect comment

**Review by minh-bq — APPROVED** [review]

**Pull request merged**